processing/src/main/java/io/druid/query/groupby/having/DimFilterHavingSpec.java

```diff
@@ -25,26 +25,29 @@
 import io.druid.data.input.Row;
 import io.druid.query.filter.DimFilter;
 import io.druid.query.filter.ValueMatcher;
-import io.druid.query.groupby.RowBasedValueMatcherFactory;
+import io.druid.query.groupby.RowBasedColumnSelectorFactory;
+import io.druid.segment.column.ValueType;
 
 import java.nio.ByteBuffer;
+import java.util.Map;
 
-public class DimFilterHavingSpec implements HavingSpec
+public class DimFilterHavingSpec extends BaseHavingSpec
 {
   private static final byte CACHE_KEY = (byte) 0x9;
 
   private final DimFilter dimFilter;
-  private final RowBasedValueMatcherFactory valueMatcherFactory;
-  private final ValueMatcher valueMatcher;
+  private final ThreadLocal<Row> rowSupplier;
+
+  private final Object valueMatcherMonitor = new Object();
+  private volatile ValueMatcher valueMatcher;
```

Comment:
Yes, `volatile` on `valueMatcher` helps with visibility of valueMatcher's contents initialized, but it doesn't (theoretically) save you from NPE. Nothing in JMM stops VM from caching/hoisting volatile field from a very long running loop, and it won't see any volatile writes to this field, because it won't re-read it.